[#1106, #1109] Add CLI flags for sshkey and token paths

Magisus opened the pull request:
This PR adds two CLI flags to the deploy command, `sshkey-path`
and `token-path`. These will be used to pass credentials to the git
provider for use when cloning the repo, overriding values from the
config for the source.

Magisus committed:
(maint) Replace deprecated Cri `required` calls

Cri has deprecated the use of `required` to denote an option with a
required arugment because it was too confusing (it does not indicate
that the option itself is required, just its arg). It is now preferred
to use `option` with the `argument: :required` keyword arg. This commit
replaces our instances of `required` with the preferred form.

Magisus committed:
[#1106, #1109] Add CLI flags for sshkey and token paths

This commit adds two CLI flags to the deploy command, `sshkey-path`
and `token-path`. These will be used to pass credentials to the git
provider for use when cloning the repo, overriding values from the
config for the source.

Magisus committed:
[#1106, #1109] Save credentials from CLI flags

This commit adds a step to the `deploy environment` command invocation
that takes the credential files as passed via the new flags, and passes
the relevant info through to the deployment object, which is responsible
for performing the bulk of the action logic.

For SSH keys, the file path is passed directly, as Rugged expects to be
passed a file path, not the key itself. Like the rest of r10k, this
still does not support password-protected keyfiles.

For OAuth, the contents of the file is read and passed through, since
Rugged expects to be given the token as a string. For `token-path`, we
also support reading from stdin when the conventional `-` is passed.

Magisus commented on lib/r10k/action/deploy/environment.rb:
```diff
@@ -43,6 +44,30 @@ def call
+            if File.exists?(path)
+              @settings[:token] = File.read(path)
+            end
+          end
```
Do we want to do validation of the file contents at this level?

mwaggett replied:
What sort of validation were you thinking? There's a certain point at which only actually attempting to clone a repo can confirm that the credentials are good, but catching other stuff sooner could be good.

Magisus replied:
I'll see if there is a spec for OAuth token format that we can rely on.

mwaggett replied:
For reference, can I see where you found what you did?

Magisus replied:
https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/ Apparently it's super open-ended besides those character restrictions.

mwaggett replied:
thanks!

justinstoller commented on lib/r10k/action/deploy/environment.rb:
```diff
@@ -32,6 +32,7 @@ def call
           @visit_ok = true
 
           expect_config!
+          extract_credentials
```
I think this should have an exclamation point since it raises.

justinstoller commented on lib/r10k/action/deploy/environment.rb:
```diff
@@ -43,6 +44,34 @@ def call
+        end
+
+        def convert_dash_to_stdin(path)
+          path == '-' ? '/dev/stdin' : path
```
Do we need to support Windows?? I don't think that `/dev/stdin` works on Windows, however I do think that reading the IO object `STDIN`/`$stdin` would be a portable way to handle that.

Magisus replied:
I think we probably do.

justinstoller commented on lib/r10k/action/deploy/environment.rb:
```diff
@@ -43,6 +44,34 @@ def call
         private
 
+        def extract_credentials
+          if @sshkey_path && @token_path
```
It's _soooo_ cool that these instance variables are defined and managed in a different file....

Magisus replied:
In a super obtuse way too. Took me a while to figure out how I thought it worked, and even then I wasn't confident till I wrote the tests >.>

mwaggett commented on spec/unit/action/deploy/environment_spec.rb:
```diff
@@ -340,6 +411,24 @@
       end
     end
+
+    describe 'with sshkey_path' do
```
(nit) should this be `sshkey-path`?

mwaggett commented on spec/unit/action/deploy/environment_spec.rb:
```diff
@@ -340,6 +411,24 @@
+
+      subject { described_class.new({ config: '/some/nonexistent/path', 'sshkey-path': '/nonexistent' }, []) }
+
+      it 'sets puppet_path' do
```
should this be `sshkey_path`?

mwaggett commented on spec/unit/action/deploy/module_spec.rb:
```diff
@@ -149,4 +157,22 @@
+
+    subject { described_class.new({ config: '/some/nonexistent/path', 'sshkey-path': '/nonexistent' }, []) }
+
+    it 'sets puppet_path' do
```
```suggestion
    it 'sets sshkey_path' do
```

mwaggett commented on lib/r10k/action/deploy/environment.rb:
```diff
@@ -43,6 +45,41 @@ def call
+            else
+              raise R10K::Error, _("{%path} does not exist, cannot load OAuth token") % { path: @token_path }
+            end
+            { token: token }
```
I assume the choice to send the token itself here, whereas we send the file path for the sshkey, is related to how we'll ultimately use them with rugged?

Magisus replied:
Yes. Rugged expects the token as a string, and a file path for the SSH key.

Magisus committed:
[#1106] Add basic OAuth token validation

This commit adds some validation to the OAuth token that is read in,
ensuring that it at least meets the character requirements for a token.

tvpartytonight commented on lib/r10k/action/deploy/environment.rb:
```diff
@@ -43,6 +45,50 @@ def call
+            end
+
+            unless valid_token?(token)
+              raise R10K::Error, _("Supplied token is not a valid OAuth token.")
```
Nit: this is a little bit vague, because it could mean the the token is expired, too long, too short, etc. I think it should probably say something invalid characters found in the token, not that the token itself is invalid.

Magisus replied:
Updated, lemme know what you think.

tvpartytonight commented on lib/r10k/action/deploy/environment.rb:
```diff
@@ -43,6 +45,50 @@ def call
+            if @sshkey_path == '-'
+              sshkey = $stdin.read
+              keypath = write_key_to_tmpfile(sshkey).path
+              { sshkey_file: keypath }
```
nit: is there a reason to switch from `@sshkey_path` to `sshkey_file`? I would just assume it should be called `sshkey_path` for consistency, but there could be reasons I'm not thinking of? I guess `sshkey_path.path` is a little hard on the 👀 ...

Magisus replied:
Sure I can change it. I didn't have a strong reason.

tvpartytonight commented on lib/r10k/action/deploy/environment.rb:
```diff
@@ -43,6 +45,50 @@ def call
+          keypath.write(sshkey)
+          keypath.close
+          keypath
+        end
```
I am vaguely worried about leaving behind keys in the Tmpfile dir, and there are some recommendations from ruby-doc about making this [more secure](https://ruby-doc.org/stdlib-2.4.0/libdoc/tempfile/rdoc/Tempfile.html#method-i-unlink-label-Unlink-before-close). 

Posting the relevant quote here:

> On POSIX systems, it’s possible to unlink a file right after creating it, and before closing it. This removes the filesystem entry without closing the file handle, so it ensures that only the processes that already had the file handle open can access the file’s contents. It’s strongly recommended that you do this if you do not want any other processes to be able to read from or write to the Tempfile, and you do not need to know the Tempfile’s filename either.

I guess the catch is Windows isn't super great about this either...but, we don't have PE on windows...yet...

Magisus replied:
I think we do need to know the name though. The thing Rugged saves and loads is a file name...

I think the ultimate solution here is to use a regular file that is encrypted with a password and has locked down perms, and then clean it up by hand when we know we're done.

Magisus replied:
I'm not sure if there's an interim thing we could be doing that might be more secure though. Thoughts, @justinstoller @pcarlisle ?

justinstoller replied:
The file should be created in a directory that is locked down and owned by the process (e.g. `~/.ssh`). I assume somewhere in `/opt/puppetlabs` will work for PE and mostlikely is the process user's homedir (if it exists) and/or the configured datadir.

I'm not sure about unlinking before closing, we will be the same process as rugged right? But not if we provide support for shellgit at some point?

As a real handwaving thing, I would look to what `puppet` uses as that code base has learned (the hard way) how to secure tmp files. [FileSystem impl](https://github.com/puppetlabs/puppet/blob/main/lib/puppet/file_system/file_impl.rb#L168-L183) & [FileSystem::UniqueFile](https://github.com/puppetlabs/puppet/blob/cd7e16542c919e516dda44dbf6bac0ff64dacf0a/lib/puppet/file_system/uniquefile.rb#L10).

justinstoller replied:
Also.... If we're writing files now, we can do that in code-manager where we know the datadir and just give the path to r10k. Right?

Magisus replied:
Yeah, that will work for our usecase, but we probably want to make sure we're still doing something sensible for FOSS users, I think? Does `~/.ssh` seem like the right call for the general case?

Magisus replied:
Or wait. Are you saying we should just only accept a filepath (and not stdin) for this? Cause that works for me too, and that puts the onus on the user to secure the key properly, which makes sense to me.

tvpartytonight replied:
I'm a little reticent to put items in `.ssh` that are ephemeral by nature and could get left behind. Do we need our own cool `.r10k-ssh` dir? 🤷

Magisus replied:
I'm also totally fine changing this to only accept a file path (since that's what we need anyway) and not messing with this at all.

justinstoller replied:
I'm thinking, the "-" stdin bit was good when our explicit goal was to not write anything to disk.

But, iiuc, we've landed on writing to disk is unavoidable.

Then for our use case we should write the secrets to disk where it makes the most sense for us. When writing to secrets to disk we should write them to a secure directory owned by the process (a directory locked down like ~/.ssh normally is). For us, I think that is our datadir in code-manager (and if not we can make a sub-directory and manage its perms).

Figuring out that where-to-write-secrets bit sounds like it would be hard from r10k (haven't done a lot of looking though, maybe I'm wrong?) and seems like we might as well write the secret to disk in the application calling r10k since it is explicitly configured with the process owned datadir.

Magisus replied:
That works for me. I'll update this. I'll leave `-` as an option for the token though, since there we really don't ever want to write it out.

justinstoller commented on lib/r10k/action/deploy/environment.rb:
```diff
@@ -43,6 +45,38 @@ def call
+        end
+
+        def valid_token?(token)
+          return token =~ /^[\w\-\.~\+\/]+$/
```
It would be valuable to add a comment pointing to your link (https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/) about why we're using this bizarre set of characters.

justinstoller commented on lib/r10k/action/deploy/environment.rb:
```diff
@@ -5,6 +5,7 @@
 require 'r10k/action/base'
 require 'r10k/action/deploy/deploy_helpers'
 require 'json'
+require 'tempfile'
```
This is unused outside of tests now, right?

Magisus replied:
Good catch.
